### PR TITLE
ProtobufDumper improvements

### PR DESCRIPTION
**Pull request by xPaw**
* Removes unsafe code and replaces it with MemoryStream.
* At least 4x speed up on `steamclient.dylib`
* Defers all protos until the end (to fix duplicate protos)

### CODE REVIEW

**Comment by xPaw** [comment]
Proof that this PR works: https://github.com/SteamDatabase/SteamTracking/commit/67f6065ff843c8784aca12d6346a49fc1877d324

**Comment by VoiDeD** [comment]
Any regressions with running this on windows binaries or is that still functional?

File: Resources/ProtobufDumper/ProtobufDumper/ImageFile.cs [hunk 1]
```diff
@@ -410,8 +405,25 @@ private void FinalPassWriteFiles()
...
                     }
                 }
-                
-                File.WriteAllText(outputFile, proto.buffer.ToString());
+
+                var buffer = proto.buffer.ToString();
+
+                if (alreadyWrittenProtos.ContainsKey(outputFile))
+                {
+                    if (alreadyWrittenProtos[outputFile] == buffer)
```
Comment (VoiDeD) [inline Resources/ProtobufDumper/ProtobufDumper/ImageFile.cs]: This is one heck of a string comparison, but it's probably fine.

File: Resources/ProtobufDumper/ProtobufDumper/ImageFile.cs [hunk 2]
```diff
@@ -668,7 +680,25 @@ private string GetValueForProp(dynamic propInfo, object options, bool field, out
                         string value = GetValueForProp(fieldInfo, deserialized, true, out name);
 
                         if (value != null)
-                            options_kv.Add(name, value);
+                        {
+                            if (options_kv.ContainsKey(name))
+                            {
+                                if (options_kv[name] == value)
+                                {
+                                    Console.WriteLine("Found a duplicate option '{0}' with same value, ignoring...", name);
```
Comment (VoiDeD) [inline Resources/ProtobufDumper/ProtobufDumper/ImageFile.cs]: I'm curious if the serialized proto descriptors actually include duplicate options in their byte steams, or if this is a bug elsewhere in protobuf dumper.

**Comment by xPaw** [comment]
Yep, Windows dumping still works.

**Comment by xPaw** [comment]
With my changes, ProtobufDumper is at about 4 times faster when run on `steamclient.dylib` file.

* master: ~3800ms
* back-to-the-future: ~930ms

**Comment by VoiDeD** [comment]
LGTM if it works.

**Comment by Netshroud** [comment]
LGTM too. No known regressions and an impressive performance boost, but I want final confirmation from AzuiSleet before merging. I'm getting mixed messages from IRC:

```
[05:28:13] <AzuiSleet> safeguard is a no-go. I'll take a look at what deferredProtos needs to do
...
[05:53:38] <AzuiSleet> alright, it looks good, closer to 4 passes though
```

(safeguard is still in there)

All OK, @azuisleet?

**Comment by Netshroud** [comment]
For the record:
```
<AzuiSleet> safeguard is OK because the previous code simply looped until 0
```

Merging...

**Pull request merged**